File: backend-flask/static/js/currency-formatter.js
/**
 * Utilitário universal para formatação monetária no frontend
 * Padrão brasileiro com símbolo R$ e separadores corretos
 * OTIMIZADO: Lazy loading e performance melhorada
 * 
 * @fileoverview Sistema universal de formatação monetária JavaScript
 * @version 1.2.0 - Corrigido (sem duplicações)
 */

// Constantes para uso em toda aplicação
if (typeof window.CURRENCY_CONSTANTS === 'undefined') {
    window.CURRENCY_CONSTANTS = {
        SYMBOL: 'R$',
        DECIMAL_PLACES: 2,
        THOUSANDS_SEPARATOR: '.',
        DECIMAL_SEPARATOR: ',',
        PATTERN: /^R?\$?\s?(\d{1,3}(\.\d{3})*),(\d{2})$/
    };
}

// Verificar se a classe já foi declarada para evitar erro de redeclaração
if (typeof window.CurrencyFormatter === 'undefined') {
    class CurrencyFormatter {
    
    /**
     * Formatar valor para moeda brasileira
     * @param {number|string|null} value - Valor a ser formatado
     * @param {boolean} includeSymbol - Se deve incluir o símbolo R$ (padrão: true)
     * @param {number} decimalPlaces - Número de casas decimais (padrão: 2)
     * @returns {string} Valor formatado no padrão brasileiro
     */
    static formatCurrency(value, includeSymbol = true, decimalPlaces = 2) {
        // Tratar valores null, undefined ou vazios
        if (value === null || value === undefined || value === '') {
            return includeSymbol ? 'R$ 0,00' : '0,00';
        }

        try {
            let numericValue;

            // Converter string para número se necessário
            if (typeof value === 'string') {
                // Remove caracteres não numéricos exceto ponto e vírgula
                let cleanValue = value.replace(/[R$\s]/g, '');
                
                // Lidar com formato brasileiro (1.234,56)
                if (cleanValue.includes(',') && cleanValue.includes('.')) {
                    const parts = cleanValue.split(',');
                    if (parts.length === 2) {
                        const integerPart = parts[0].replace(/\./g, '');
                        const decimalPart = parts[1];
                        cleanValue = `${integerPart}.${decimalPart}`;
                    }
                } else if (cleanValue.includes(',')) {
                    // Apenas vírgula, substituir por ponto
                    cleanValue = cleanValue.replace(',', '.');
                }

                numericValue = parseFloat(cleanValue);
            } else {
                numericValue = parseFloat(value);
            }

            // Verificar se é um número válido
            if (isNaN(numericValue)) {
                return includeSymbol ? 'R$ 0,00' : '0,00';
            }

            // Arredondar para o número de casas decimais especificado
            const rounded = Math.round(numericValue * Math.pow(10, decimalPlaces)) / Math.pow(10, decimalPlaces);

            // Separar parte inteira e decimal
            const isNegative = rounded < 0;
            const absoluteValue = Math.abs(rounded);
            const integerPart = Math.floor(absoluteValue);
            const decimalPart = (absoluteValue - integerPart).toFixed(decimalPlaces).substring(2);

            // Formatar parte inteira com pontos como separadores de milhares
            const formattedInteger = integerPart.toString().replace(/\B(?=(\d{3})+(?!\d))/g, '.');

            // Combinar partes
            let formattedValue = `${formattedInteger},${decimalPart}`;

            // Adicionar sinal negativo se necessário
            if (isNegative) {
                formattedValue = `-${formattedValue}`;
            }

            // Adicionar símbolo R$ se solicitado
            return includeSymbol ? `R$ ${formattedValue}` : formattedValue;

        } catch (error) {
            console.warn('Erro na formatação monetária:', error);
            return includeSymbol ? 'R$ 0,00' : '0,00';
        }
    }

    /**
     * Converter string formatada em moeda para número
     * @param {string} formattedValue - String no formato "R$ 1.234,56"
     * @returns {number} Valor numérico correspondente
     */
    static parseCurrency(formattedValue) {
        if (!formattedValue || typeof formattedValue !== 'string') {
            return 0;
        }

        try {
            // Remove R$ e espaços
            let cleanValue = formattedValue.replace(/[R$\s]/g, '');

            // Lidar com formato brasileiro
            if (cleanValue.includes(',') && cleanValue.includes('.')) {
                const parts = cleanValue.split(',');
                if (parts.length === 2) {
                    const integerPart = parts[0].replace(/\./g, '');
                    const decimalPart = parts[1];
                    cleanValue = `${integerPart}.${decimalPart}`;
                }
            } else if (cleanValue.includes(',')) {
                cleanValue = cleanValue.replace(',', '.');
            }

            return parseFloat(cleanValue) || 0;

        } catch (error) {
            console.warn('Erro no parsing monetário:', error);
            return 0;
        }
    }

    /**
     * Validar entrada de valor monetário
     * @param {string} value - Valor a ser validado
     * @returns {object} {isValid: boolean, parsedValue: number}
     */
    static validateCurrencyInput(value) {
        try {
            const parsed = this.parseCurrency(value);
            return {
                isValid: !isNaN(parsed) && isFinite(parsed),
                parsedValue: parsed
            };
        } catch (error) {
            return {
                isValid: false,
                parsedValue: 0
            };
        }
    }

    /**
     * Formatar automaticamente input de moeda durante digitação
     * @param {HTMLInputElement} input - Elemento input
     */
    static setupCurrencyInput(input) {
        if (!input || !input.addEventListener) {
            console.warn('Elemento input inválido para formatação monetária');
            return;
        }

        // Evento de input para formatação em tempo real
        input.addEventListener('input', function(e) {
            const cursorPosition = e.target.selectionStart;
            const value = e.target.value;
            const numericValue = window.CurrencyFormatter.parseCurrency(value);
            const formattedValue = window.CurrencyFormatter.formatCurrency(numericValue, false);
            
            e.target.value = formattedValue;
            
            // Manter posição do cursor aproximada
            const newCursorPosition = Math.min(cursorPosition, formattedValue.length);
            e.target.setSelectionRange(newCursorPosition, newCursorPosition);
        });

        // Evento de blur para garantir formatação final
        input.addEventListener('blur', function(e) {
            const numericValue = window.CurrencyFormatter.parseCurrency(e.target.value);
            e.target.value = window.CurrencyFormatter.formatCurrency(numericValue, false);
        });

        // Permitir apenas números, vírgula, ponto e teclas de controle
        input.addEventListener('keypress', function(e) {
            const char = String.fromCharCode(e.which);
            if (!/[\d,.]/.test(char) && 
                !['Backspace', 'Delete', 'Tab', 'Enter', 'ArrowLeft', 'ArrowRight'].includes(e.key)) {
                e.preventDefault();
            }
        });
    }

    /**
     * Aplicar formatação monetária a todos os elementos com classe específica
     * @param {string} className - Nome da classe CSS (padrão: 'currency-input')
     */
    static initializeAllCurrencyInputs(className = 'currency-input') {
        const inputs = document.querySelectorAll(`.${className}`);
        inputs.forEach(input => {
            this.setupCurrencyInput(input);
        });
    }

    /**
     * Formatar todos os elementos de texto com valores monetários
     * @param {string} selector - Seletor CSS (padrão: '.currency-display')
     */
    static formatAllCurrencyDisplays(selector = '.currency-display') {
        const elements = document.querySelectorAll(selector);
        elements.forEach(element => {
            const value = element.textContent || element.innerText;
            const formatted = this.formatCurrency(value);
            element.textContent = formatted;
            element.setAttribute('data-original-value', value);
        });
    }
}

    // Registrar na janela global
    window.CurrencyFormatter = CurrencyFormatter;
}

// Funções globais para compatibilidade e facilidade de uso
if (typeof window.formatCurrency === 'undefined') {
    window.formatCurrency = (value, includeSymbol, decimalPlaces) => 
        window.CurrencyFormatter.formatCurrency(value, includeSymbol, decimalPlaces);

    window.parseCurrency = (value) => 
        window.CurrencyFormatter.parseCurrency(value);

    window.formatMoeda = (value) => 
        window.CurrencyFormatter.formatCurrency(value);

    window.formatReal = (value) => 
        window.CurrencyFormatter.formatCurrency(value);
}

// Inicialização automática quando o DOM estiver pronto
document.addEventListener('DOMContentLoaded', function() {
    if (typeof window.CurrencyFormatter !== 'undefined') {
        // Inicializar inputs de moeda automaticamente
        window.CurrencyFormatter.initializeAllCurrencyInputs();
        
        // Formatar displays de moeda automaticamente
        window.CurrencyFormatter.formatAllCurrencyDisplays();
        
        console.log('✅ Sistema de formatação monetária inicializado');
    }
});

// Exportar para uso em módulos
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        CurrencyFormatter: window.CurrencyFormatter,
        formatCurrency: window.CurrencyFormatter.formatCurrency,
        parseCurrency: window.CurrencyFormatter.parseCurrency,
        CURRENCY_CONSTANTS: window.CURRENCY_CONSTANTS
    };
}

File: backend-flask/utils/currency_formatter.py
"""
Utilitário universal para formatação monetária
Padrão brasileiro com símbolo R$ e separadores corretos
"""

import locale
from decimal import Decimal, ROUND_HALF_UP
from typing import Union, Optional

def format_currency(value: Union[float, int, str, Decimal, None], 
                   include_symbol: bool = True, 
                   decimal_places: int = 2) -> str:
    """
    Formata valor para moeda brasileira com padrão R$ XX.XXX,XX
    
    Args:
        value: Valor a ser formatado (float, int, str, Decimal ou None)
        include_symbol: Se deve incluir o símbolo R$ (padrão: True)
        decimal_places: Número de casas decimais (padrão: 2)
    
    Returns:
        String formatada no padrão brasileiro
        
    Examples:
        format_currency(1234.56) -> "R$ 1.234,56"
        format_currency(1234.56, False) -> "1.234,56"
        format_currency(0) -> "R$ 0,00"
        format_currency(None) -> "R$ 0,00"
    """
    
    # Tratar valores None ou vazios
    if value is None or value == "":
        return "R$ 0,00" if include_symbol else "0,00"
    
    try:
        # Converter para Decimal para precisão
        if isinstance(value, str):
            # Remove caracteres não numéricos exceto ponto e vírgula
            clean_value = value.replace("R$", "").replace(" ", "").strip()
            # Substituir vírgula por ponto se necessário
            if "," in clean_value and "." not in clean_value:
                clean_value = clean_value.replace(",", ".")
            elif "," in clean_value and "." in clean_value:
                # Formato brasileiro: 1.234,56 -> 1234.56
                parts = clean_value.split(",")
                if len(parts) == 2:
                    integer_part = parts[0].replace(".", "")
                    decimal_part = parts[1]
                    clean_value = f"{integer_part}.{decimal_part}"
            value = float(clean_value)
        
        # Converter para Decimal
        decimal_value = Decimal(str(value))
        
        # Arredondar para o número de casas decimais especificado
        rounded_value = decimal_value.quantize(
            Decimal('0.' + '0' * decimal_places), 
            rounding=ROUND_HALF_UP
        )
        
        # Converter para float para formatação
        float_value = float(rounded_value)
        
        # Separar parte inteira e decimal
        integer_part = int(abs(float_value))
        decimal_part = abs(float_value) - integer_part
        
        # Formatar parte inteira com pontos como separadores de milhares
        formatted_integer = f"{integer_part:,}".replace(",", ".")
        
        # Formatar parte decimal
        formatted_decimal = f"{decimal_part:.{decimal_places}f}"[2:]  # Remove "0."
        
        # Combinar partes
        formatted_value = f"{formatted_integer},{formatted_decimal}"
        
        # Adicionar sinal negativo se necessário
        if float_value < 0:
            formatted_value = f"-{formatted_value}"
        
        # Adicionar símbolo R$ se solicitado
        if include_symbol:
            return f"R$ {formatted_value}"
        else:
            return formatted_value
            
    except (ValueError, TypeError, AttributeError):
        # Em caso de erro, retornar valor padrão
        return "R$ 0,00" if include_symbol else "0,00"


def parse_currency(formatted_value: str) -> float:
    """
    Converte string formatada em moeda para float
    
    Args:
        formatted_value: String no formato "R$ 1.234,56" ou "1.234,56"
    
    Returns:
        Valor float correspondente
        
    Examples:
        parse_currency("R$ 1.234,56") -> 1234.56
        parse_currency("1.234,56") -> 1234.56
    """
    
    if not formatted_value or formatted_value.strip() == "":
        return 0.0
    
    try:
        # Remove R$ e espaços
        clean_value = str(formatted_value).replace("R$", "").strip()
        
        # Se contém vírgula e ponto, é formato brasileiro
        if "," in clean_value and "." in clean_value:
            parts = clean_value.split(",")
            if len(parts) == 2:
                integer_part = parts[0].replace(".", "")
                decimal_part = parts[1]
                return float(f"{integer_part}.{decimal_part}")
        
        # Se contém apenas vírgula, substituir por ponto
        elif "," in clean_value:
            clean_value = clean_value.replace(",", ".")
        
        # Se contém apenas pontos, verificar se é separador de milhares ou decimal
        elif "." in clean_value:
            parts = clean_value.split(".")
            if len(parts) > 2:
                # Múltiplos pontos = separadores de milhares + decimal
                integer_parts = parts[:-1]
                decimal_part = parts[-1]
                integer_value = "".join(integer_parts)
                clean_value = f"{integer_value}.{decimal_part}"
        
        return float(clean_value)
        
    except (ValueError, TypeError, AttributeError):
        return 0.0


# Função para uso direto em templates Jinja2
def currency_filter(value):
    """Filtro Jinja2 para formatação monetária"""
    return format_currency(value)


# Função para validação de entrada monetária
def validate_currency_input(value: str) -> tuple[bool, float]:
    """
    Valida entrada de valor monetário
    
    Returns:
        tuple: (is_valid, parsed_value)
    """
    try:
        parsed = parse_currency(value)
        return True, parsed
    except:
        return False, 0.0


# Constantes para uso em toda aplicação
CURRENCY_SYMBOL = "R$"
CURRENCY_DECIMAL_PLACES = 2
CURRENCY_THOUSANDS_SEPARATOR = "."
CURRENCY_DECIMAL_SEPARATOR = ","